Library/OcCryptoLib/RsaDigitalSign.c

```diff
@@ -0,0 +1,511 @@
+  B = Mula32 (D0, Key->N[0], (UINT32) A);
+
+  for (Index = 1; Index < Key->Size; Index++) {
+    A = Mulaa32 (Aa, Bb[Index], C[Index], (UINT32) (A >> 32));
+    B = Mulaa32 (D0, Key->N[Index], (UINT32) A, (UINT32) (B >> 32));
+    C[Index - 1] = (UINT32) B;
+  }
+
+  A = (A >> 32) + (B >> 32);
+  C[Index - 1] = (UINT32) A;
+
+  if (A >> 32) {
+    SubMod (Key, C);
+  }
+}
+
+//
+// Montgomery c[] = a[] * b[] / R % mod
+//
+STATIC
+VOID
+MontMul (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *C,
+  UINT32          *A,
+  UINT32          *B
+  )
+{
+  UINT32 Index;
+
+  ZeroMem (C, Key->Size);
+
+  for (Index = 0; Index < Key->Size; Index++) {
+    MontMulAdd (Key, C, A[Index], B);
+  }
+}
+
+/**
+  In-place public exponentiation.
+  Exponent depends on the configuration (65537 (default), or 3).
+
+  @param Key        Key to use in signing
+  @param InOut      Input and output big-endian byte array
+ **/
+STATIC
+VOID
+ModPow (
+  RSA_PUBLIC_KEY  *Key,
+  UINT8           *InOut
+  )
+{
+  UINT32 *A     = NULL;
+  UINT32 *Ar    = NULL;
+  UINT32 *Aar   = NULL;
+  UINT32 *Aaa   = NULL;
+  INT32  Index  = 0;
+  UINT32 Tmp    = 0;
+
+  A = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  Ar = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  Aar = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  if (A == NULL || Ar == NULL || Aar == NULL) {
+    if (A != NULL) {
+      FreePool (A);
+    }
+    if (Ar != NULL) {
+      FreePool (Ar);
+    }
+    if (Aar != NULL) {
+      FreePool (Aar);
+      // TODO: debug_error about efi memory alocation
+    }
+    return;
+  }
+
+  //
+  // Re-use location
+  //
+  Aaa = Aar;
+
+  //
+  // Convert from big endian byte array to little endian word array
+  //
+  for (Index = 0; Index < (INT32) Key->Size; Index++) {
+    Tmp =
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 0] << 24) |
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 1] << 16) |
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 2] << 8) |
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 3] << 0);
+    A[Index] = Tmp;
+  }
+
+  //
+  // Ar = A * Rr / R mod M
+  //
+  MontMul (Key, Ar, A, Key->Rr);
+  //
+  // Exponent 65537
+  //
+  for (Index = 0; Index < 16; Index += 2) {
+    //
+    // Aar = Ar * Ar / R mod M 
+    //
+    MontMul (Key, Aar, Ar, Ar);
+    //
+    // Ar = Aar * Aar / R mod M 
+    //
+    MontMul (Key, Ar, Aar, Aar);
+  }
+  //
+  // Aaa = Ar * A / R mod M
+  //
+  MontMul (Key, Aaa, Ar, A);
+
+  //
+  // Make sure Aaa < Mod; Aaa is at most 1x mod too large.
+  //
+  if (GeMod (Key, Aaa)){
+    SubMod (Key, Aaa);
+  }
+
+  //
+  // Convert to bigendian byte array
+  //
+  for (Index = (INT32) Key->Size - 1; Index >= 0; --Index) {
+    Tmp = Aaa[Index];
+
+    *InOut++ = (UINT8) (Tmp >> 24);
+    *InOut++ = (UINT8) (Tmp >> 16);
+    *InOut++ = (UINT8) (Tmp >>  8);
+    *InOut++ = (UINT8) (Tmp >>  0);
+  }
+
+  //
+  // Free work buffers before return
+  //
+  if (A != NULL) {
```

Comment:
Checking for non NULL makes no sense here, it was done above.

Reply:
Done. https://github.com/acidanthera/OcSupportPkg/commit/2b45105bd38b96bd375e62c37630dfa377b3bc61